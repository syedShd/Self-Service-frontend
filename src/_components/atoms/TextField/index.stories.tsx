import { Story, Meta } from '@storybook/react'
import { ThemeProvider } from '@material-ui/core/styles'
import VisibilityIcon from '@material-ui/icons/Visibility'
import TextField from '.'
import baseTheme from '../../../core-utils/theme'

export default {
  title: 'atoms/TextField',
  component: TextField,
  argTypes: {
    variant: {
      options: ['filled', 'outlined', 'standard'],
      control: { type: 'select' },
    },
  },
} as Meta

const Template: Story = ({ value, ...args }) => (
  <ThemeProvider theme={baseTheme}>
    <TextField value={value} {...args} />
  </ThemeProvider>
)

export const WithLabel = Template.bind({})
WithLabel.args = {
  variant: 'outlined',
  value: 'Enter Text',
  label: 'Label',
}

export const WithoutLabel = Template.bind({})
WithoutLabel.args = {
  variant: 'outlined',
  value: 'Enter Text',
}
export const WithPlaceHolder = Template.bind({})
WithPlaceHolder.args = {
  variant: 'outlined',
  placeholder: 'Enter Text',
  label: 'Label',
}

export const WithError = Template.bind({})
WithError.args = {
  variant: 'outlined',
  value: 'some error',
  label: 'Label',
  error: true,
}

export const FilledTextField = Template.bind({})
FilledTextField.args = {
  variant: 'filled',
  value: 'Enter Text',
  label: 'Label',
}
export const FullWidthTextField = Template.bind({})
FullWidthTextField.args = {
  variant: 'outlined',
  value: 'Enter Text',
  label: 'Label',
  fullWidth: true,
}

export const WithendAdormentIcon = Template.bind({})
WithendAdormentIcon.args = {
  variant: 'outlined',
  value: 'Enter Text',
  label: 'Label',
  endAdorment: <VisibilityIcon />,
}

export const WithendAdorment = Template.bind({})
WithendAdorment.args = {
  variant: 'outlined',
  value: '23',
  label: 'Label',
  endAdorment: 'kg',
}
