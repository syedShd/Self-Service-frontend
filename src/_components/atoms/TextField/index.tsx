import React from 'react'
import PropTypes from 'prop-types'
import { TextField as MUITextField, InputAdornment } from '@material-ui/core'
import { makeStyles } from '@material-ui/core/styles'
import Typography from '../Typography'
import { textFieldmargin, textFieldVariants } from '../../../core-utils/helper'

const useStyles = makeStyles((theme) => ({
  root: {
    '& .MuiFilledInput-root': {
      background: 'rgb(232, 241, 250)',
    },
  },
  helperText: {
    marginLeft: 0,
    marginRight: 0,
    fontSize: 14,
    lineHeight: '21px',
    letterSpacing: '0.25px',
  },
  label: {
    ...theme.typography.body2,
    color: theme.palette.info.main,
  },
}))

type TextFieldProps = {
  label?: string
  name?: string
  value?: string
  multiline?: boolean
  rows?: number
  rowsMax?: number
  placeholder?: string
  handleChange?: ((value: string) => void) | null
  variant?: string
  margin?: string
  helperText?: string
  [restProps: string]: unknown
  fullWidth?: boolean
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  endAdorment?: any
}

const TextField = ({
  label,
  name,
  value,
  multiline,
  rows,
  rowsMax,
  placeholder,
  handleChange,
  variant,
  margin,
  helperText,
  endAdorment,
  ...restProps
}: TextFieldProps) => {
  const classes = useStyles()

  const onChange = ({
    target: { value: currentValue },
  }: React.ChangeEvent<HTMLInputElement>) => {
    handleChange?.(currentValue)
  }

  return (
    <>
      {label && (
        <Typography data-testid="textfield-label" className={classes.label}>
          {label ?? ''}
        </Typography>
      )}
      <MUITextField
        InputProps={{
          endAdornment: (
            <InputAdornment position="end">{endAdorment}</InputAdornment>
          ),
        }}
        name={name}
        margin={textFieldmargin(margin ?? '')}
        value={value}
        multiline={multiline}
        placeholder={placeholder}
        rows={rows}
        variant={textFieldVariants(variant ?? '')}
        rowsMax={rowsMax}
        onChange={onChange}
        helperText={helperText}
        fullWidth
        {...restProps}
      />
    </>
  )
}

TextField.propTypes = {
  label: PropTypes.string,
  name: PropTypes.string,
  value: PropTypes.string,
  multiline: PropTypes.bool,
  rows: PropTypes.number,
  rowsMax: PropTypes.number,
  placeholder: PropTypes.string,
  handleChange: PropTypes.func,
  variant: PropTypes.string,
  margin: PropTypes.string,
  helperText: PropTypes.string,
  fullWidth: PropTypes.bool,
}

TextField.defaultProps = {
  label: '',
  name: '',
  multiline: false,
  rows: 1,
  rowsMax: 1,
  placeholder: '',
  handleChange: null,
  variant: 'outlined',
  margin: 'none',
  helperText: '',
  value: '',
  fullWidth: false,
  endAdorment: null,
}

export default TextField
