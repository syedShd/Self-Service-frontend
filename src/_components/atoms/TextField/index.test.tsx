import React from 'react'
import MUITextField from '@material-ui/core/TextField'
import { mount, shallow } from 'enzyme'
import TextField from '.'

describe('Text Field  Component', () => {
  it('Renders correctly', () => {
    const wrapper = shallow(<TextField />)
    expect(wrapper).toMatchSnapshot()
  })

  it('Allows us to set props', () => {
    const wrapper = mount(
      <TextField
        label="test"
        placeholder="test placeholder"
        endAdorment="kg"
        margin="dense"
        variant="filled"
      />
    )
    expect(wrapper.props().label).toEqual('test')
    expect(wrapper.props().placeholder).toEqual('test placeholder')
    wrapper.setProps({ label: 'foo' })
    wrapper.setProps({ placeholder: 'foo placeholder' })
    expect(wrapper.props().placeholder).toEqual('foo placeholder')
    expect(wrapper.props().label).toEqual('foo')
  })

  it('onChange works', () => {
    const mockHandleChangeText = jest.fn()
    const component = mount(
      <TextField
        name="field"
        onChange={mockHandleChangeText}
        endAdorment="kg"
        margin="normal"
        variant="standard"
      />
    )
    component
      ?.find(MUITextField)
      ?.props()
      ?.onChange?.({ target: { value: 'test' } } as any)
    expect(mockHandleChangeText).toHaveBeenCalledWith({
      target: { value: 'test' },
    })
  })
})
