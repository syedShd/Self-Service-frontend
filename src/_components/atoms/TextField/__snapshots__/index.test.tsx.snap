// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text Field  Component Renders correctly 1`] = `
<Fragment>
  <WithStyles(ForwardRef(TextField))
    InputProps={
      Object {
        "endAdornment": <WithStyles(ForwardRef(InputAdornment))
          position="end"
        />,
      }
    }
    fullWidth={false}
    helperText=""
    margin="none"
    multiline={false}
    name=""
    onChange={[Function]}
    placeholder=""
    rows={1}
    rowsMax={1}
    value=""
    variant="outlined"
  />
</Fragment>
`;
